src/main/scala/com/gu/elasticsearchmonitor/MasterDetector.scala

```diff
@@ -0,0 +1,64 @@
+package com.gu.elasticsearchmonitor
+
+import com.amazonaws.services.ec2.AmazonEC2
+import com.amazonaws.services.ec2.model.{ DescribeInstancesRequest, Filter, Instance }
+import okhttp3.{ OkHttpClient, Request }
+import org.slf4j.{ Logger, LoggerFactory }
+
+import collection.JavaConverters._
+import scala.annotation.tailrec
+import scala.util.{ Failure, Random, Success, Try }
+
+class MasterDetector(amazonEC2: AmazonEC2, httpClient: OkHttpClient) {
+
+  val logger: Logger = LoggerFactory.getLogger(this.getClass)
+
+  def detectMasters(env: Env): Either[String, MasterInformation] = {
+    @tailrec
+    def queryInstances(instancesFromPreviousRequest: List[Instance], nextToken: Option[String]): List[Instance] = {
+      val filters = List(
+        new Filter("tag:Stage", List(env.stage).asJava),
+        new Filter("tag:Stack", List(env.tagQueryStack).asJava),
+        new Filter("tag:App", List(env.tagQueryApp).asJava))
+      val dir = new DescribeInstancesRequest().withFilters(filters: _*).withNextToken(nextToken.orNull)
+      val result = amazonEC2.describeInstances(dir)
+      val instances = instancesFromPreviousRequest ++ result.getReservations.asScala.flatMap(_.getInstances.asScala)
+      if (result.getNextToken == null) {
+        instances
+      } else {
+        queryInstances(instances, Option(result.getNextToken))
+      }
+    }
+
+    def pingInstance(instanceName: String): Boolean = {
+      val clusterHealthRequest = new Request.Builder()
+        .url(s"$instanceName/_cluster/health")
+        .build()
+
+      val clusterHealthResponse = Try(httpClient.newCall(clusterHealthRequest).execute())
+      clusterHealthResponse match {
+        case Success(response) if response.code == 200 => true
+        case _ => false
+      }
+    }
+
+    Try(queryInstances(Nil, None)) match {
+      case Success(instances) =>
+        val instanceNames = instances.map(instance => s"http://${instance.getPrivateIpAddress}:9200")
+        logger.info(s"Identified these instances as potential candidates: $instanceNames")
+        val statuses = instanceNames.map(pingInstance)
+
+        val aliveInstances = instanceNames.zip(statuses).filter(_._2).map(_._1)
```

Comment:
Minor: rather than needing to `zip` here, could we construct the tuples on line 49? We already have each instanceName so we could pair it up with the result from pingInstance. 

This would make statuses a `List[(String, Boolean)]`

Reply:
you're right I didn't like that `.filter(_._2).map(_._1)`

I've replaced it with a filter which I think is much nicer